src/agg_bookend_optimize.c

```diff
@@ -0,0 +1,678 @@
+}
+
+/*
+ * replace_aggref_in_tlist
+ *
+ * If MinMaxAggPath is chosen, instead of running aggregate
+ * function we will execute subquery that we've generated. Since we
+ * use subquery we need to replace target list Aggref node with Param
+ * node. Param node passes output value from the subquery.
+ *
+ */
+void
+replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path)
+{
+	MutatorContext context;
+
+	context.mm_path = minmaxagg_path;
+
+	((Path *) minmaxagg_path)->pathtarget->exprs = (List *) mutate_aggref_node((Node *) ((Path *) minmaxagg_path)->pathtarget->exprs, (void *) &context);
+}
+
+static bool
+is_first_last_agg(Aggref *aggref, const char *name)
+{
+	char	   *agg_name = get_func_name(aggref->aggfnoid);
+
+	if (strncmp(agg_name, name, NAMEDATALEN) == 0)
+		return true;
+	return false;
+}
+
+static bool
+is_first_agg(Aggref *aggref)
+{
+	return is_first_last_agg(aggref, FIRST);
+}
+
+static bool
+is_last_agg(Aggref *aggref)
+{
+	return is_first_last_agg(aggref, LAST);
+}
+
+static bool
+contains_aggref_node(List *sortClause, List *targetList)
+{
+	List	   *exprs = get_sortgrouplist_exprs(sortClause, targetList);
+	ListCell   *cell;
+
+	foreach(cell, exprs)
+	{
+		Node	   *expr = lfirst(cell);
+
+		if (IsA(expr, Aggref))
+		{
+			Aggref	   *aggref = (Aggref *) expr;
+
+			if (is_first_agg(aggref) || is_last_agg(aggref))
+				return true;
+		}
+	}
+	return false;
+}
+
+/*
+ * preprocess_first_last_aggregates - preprocess FIRST/LAST aggregates
+ *
+ * Check to see whether the query contains FIRST/LAST aggregate functions that
+ * might be optimizable via index scans.  If it does, and all the aggregates
+ * are potentially optimizable, then create a MinMaxAggPath(reusing MinMax path implementation)\
+ * and add it to the (UPPERREL_GROUP_AGG, NULL) upperrel.
+ *
+ * This should be called by grouping_planner() just before it's ready to call
+ * query_planner(), because we generate indexscan paths by cloning the
+ * planner's state and invoking query_planner() on a modified version of
+ * the query parsetree.  Thus, all preprocessing needed before query_planner()
+ * must already be done.
+ *
+ * Note: we are passed the preprocessed targetlist separately, because it's
+ * not necessarily equal to root->parse->targetList.
+ */
+void
+preprocess_first_last_aggregates(PlannerInfo *root, List *tlist)
+{
+	Query	   *parse = root->parse;
+	FromExpr   *jtnode;
+	RangeTblRef *rtr;
+	RangeTblEntry *rte;
+	List	   *first_last_aggs;
+	RelOptInfo *grouped_rel;
+	ListCell   *lc;
+	List	   *mm_agg_list;
+	MinMaxAggPath *minmaxagg_path;
+
+	/* minmax_aggs list should be empty at this point */
+	Assert(root->minmax_aggs == NIL);
+
+	/* Nothing to do if query has no aggregates */
+	if (!parse->hasAggs)
+		return;
+
+	Assert(!parse->setOperations);	/* shouldn't get here if a setop */
+	Assert(parse->rowMarks == NIL); /* nor if FOR UPDATE */
+
+	/*
+	 * Reject unoptimizable cases.
+	 *
+	 * We don't handle the case when agg function is in ORDER BY.
```

Comment:
Can you expand on why and how we could in the future?

Reply:
I've added explanation. I don't think it's a good practice to provide future plans as code comments ( since it might never happen or it happens in completely different way)